File: src/renderer/db/cjkrad.ts
import db from "./data";

interface ICjkrad {
    sub: Promise<string[]>;
    super: Promise<string[]>;
    variant: Promise<string[]>;
}

function _getComponent(c: string, part: string): Promise<string[]> {
    return new Promise((resolve, reject) => {
        db.all(`
        SELECT s.character AS x
        FROM character AS c
        INNER JOIN ${part} AS cs ON cs.character_id = c.id
        INNER JOIN character AS s ON cs.${part}_id = s.id
        WHERE c.character = ?
        ORDER BY s.frequency ASC`, [c], (err, rows) => { err ? reject(err) : resolve(rows.map((el) => el.x)); });
    });
}

export default function getRadical(c: string): ICjkrad {
    return {
        sub: _getComponent(c, "sub"),
        super: _getComponent(c, "super"),
        variant: _getComponent(c, "variant")
    };
}

File: kotlin/src/main/kotlin/hanzileveljs/Cjkrad.kt
package hanzileveljs

class Cjkrad {
    data class CjkradEntry internal constructor(
            val sub: List<String>,
            val `super`: List<String>,
            val variant: List<String>
    )

    private data class Entry (val x: String)

    private fun getComponent(s: String, part: String): List<String> {
        return Config.db.open().createQuery("""
            SELECT s.entry AS x
            FROM token AS c
            INNER JOIN $part AS p ON p.token_id = c.id
            INNER JOIN token AS s ON p.${part}_id = s.id
            WHERE c.entry = :s
            ORDER BY s.frequency ASC
        """.trimIndent())
                .addParameter("s", s)
                .executeAndFetch(Entry::class.java)
                .map { it.x }
    }

    operator fun get(s: String): CjkradEntry {
        return CjkradEntry(
                sub = getComponent(s, "sub"),
                `super` = getComponent(s, "super"),
                variant = getComponent(s, "variant")
        )
    }
}